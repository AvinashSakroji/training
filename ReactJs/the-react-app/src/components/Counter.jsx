import React, { Component } from 'react'
import { AppContext } from '../context/AppContext'
import SimpleHOC from './SimpleHOC';

class Counter extends Component {

    state = {
        count: 10
    }
    inputRef = null;

    constructor(props) {
        super(props);
        // It is compulsory to bind the method,
        // It will bind the class with the function.
        this.inc = this.inc.bind(this);
        this.decr = this.decr.bind(this);
    }

    inc() {
        //this.state.count++; //this will not work
        const updatedCount = this.state.count + 1;
        //this.setState(newState, callback_function)
        //it is invoked Async
        this.setState({
            count: updatedCount
        }, () => {
            console.log("state Count", this.state.count)
        })
    }

    decr() {
        //this.state.count++; //this will not work
        const updatedCount = this.state.count - 1;
        //this.setState(newState, callback_function)
        //it is invoked Async
        this.setState({
            count: updatedCount
        }, () => {
            console.log("state Count", this.state.count)
        })
    }

    dec = () => {
        const updatedCount = this.state.count - 1;
        this.setState({
            count: updatedCount
        }, () => {
            console.log("state Count", this.state.count)
        })
    }

    change = (e) => {
        this.setState({
            count: e.target.value
        })
    }

    update = (e) => {
        this.setState({
            count: this.inputRef.value ? parseInt(this.inputRef.value) :
                this.inputRef.value
        })
    }
    render() {
        return (
            <div>
                <h2>Counter : {this.props.title} -- {this.state.count}</h2>
                <p>This is a class based Component</p>

                <div>
                    <p> App Name : {this.context.appName}</p>
                    <p> User Name : {this.context.userName}</p>
                </div>

                <div>
                    <button id= "incBtn" onClick={this.inc}>Increment</button> &nbsp;
                    <button onClick={this.decr}>Decrement</button>
                </div>
                <div>
                    Count: <input type="number" value={this.state.count} onChange={this.change} />
                </div>

                <div>
                    Count :<input type="number" ref={r => this.inputRef = r} /> &nbsp;
                    <button onClick={this.update}>Update</button>
                </div>
            </div>
        );
    }

}
Counter.contextType = AppContext;

export default Counter;