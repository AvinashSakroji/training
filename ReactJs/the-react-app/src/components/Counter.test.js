import React from 'react'
import Counter from './Counter'

import {mount, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16'
import {create} from 'react-test-renderer'

configure({adapter: new Adapter()});

describe("Counter Tests",()=>{

    it("should be created",()=>{
        const container = mount(<Counter/>);
        const instance = container.instance();
        expect(instance).toBeTruthy();
    });

    it("should have state count as 10",()=>{
        const container = mount(<Counter/>);
        const state = container.state();
        expect(state.count).toBe(10);
    });

    it("should increment count by 1",()=>{
        const container = mount(<Counter/>);  
        const instance = container.instance();
        instance.inc();

        const state = container.state();
        expect(state.count).toBe(11);
    });

    it("should increment count by 1 by button click",()=>{
        const container = mount(<Counter/>); 
        const button = container.find("#incBtn");
        button.simulate('click');
        button.simulate('click');
        button.simulate('click');
        
        const state = container.state();
        expect(state.count).toBe(13);
    });

    it("snapshot",()=>{

        const container = create(<Counter/>);
        expect(container.toJSON()).toMatchSnapshot();

    })
});