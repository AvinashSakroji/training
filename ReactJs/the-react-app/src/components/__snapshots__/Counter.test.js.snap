// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Tests snapshot 1`] = `
<div>
  <h2>
    Counter : 
     -- 
    10
  </h2>
  <p>
    This is a class based Component
  </p>
  <div>
    <p>
       App Name : 
      Training
    </p>
    <p>
       User Name : 
      Anil
    </p>
  </div>
  <div>
    <button
      id="incBtn"
      onClick={[Function]}
    >
      Increment
    </button>
      
    <button
      onClick={[Function]}
    >
      Decrement
    </button>
  </div>
  <div>
    Count: 
    <input
      onChange={[Function]}
      type="number"
      value={10}
    />
  </div>
  <div>
    Count :
    <input
      type="number"
    />
      
    <button
      onClick={[Function]}
    >
      Update
    </button>
  </div>
</div>
`;
